src/Veldrid.Utilities/ObjParser.cs

```diff
@@ -265,82 +324,105 @@ public ObjFile FinalizeFile()
-                }
+                throw CreateParseException(location, new FormatException());
             }
 
-            private Vector2 ParseVector2(string xStr, string yStr, string location)
+            private Vector2 ParseVector2(ReadOnlySpan<char> xStr, ReadOnlySpan<char> yStr, string location)
             {
-                try
-                {
-                    float x = float.Parse(xStr, CultureInfo.InvariantCulture);
-                    float y = float.Parse(yStr, CultureInfo.InvariantCulture);
+                if (FastParse.TryParseDouble(xStr, out double x) &&
+                    FastParse.TryParseDouble(yStr, out double y))
+                    return new Vector2((float)x, (float)y);
 
-                    return new Vector2(x, y);
-                }
-                catch (FormatException fe)
-                {
-                    throw CreateParseException(location, fe);
-                }
+                throw CreateParseException(location, new FormatException());
             }
 
-            private int ParseInt(string intStr, string location)
+            private int ParseInt(ReadOnlySpan<char> intStr, string location)
             {
-                try
-                {
-                    int i = int.Parse(intStr, CultureInfo.InvariantCulture);
-                    return i;
-                }
-                catch (FormatException fe)
-                {
-                    throw CreateParseException(location, fe);
-                }
+                if (FastParse.TryParseDouble(intStr, out double result, out bool hasFraction) && !hasFraction)
+                    return (int)result;
+
+                throw CreateParseException(location, new FormatException());
             }
 
-            private void ExpectExactly(string[] pieces, int count, string name)
+            private void ExpectPieces(
+                ref ReadOnlySpanSplitter<char> pieces, string name, bool exact,
+                out ReadOnlySpan<char> piece0, out ReadOnlySpan<char> piece1, out ReadOnlySpan<char> piece2)
             {
-                if (pieces.Length != count + 1)
-                {
-                    string message = string.Format(
-                        "Expected exactly {0} components to a line starting with {1}, on line {2}, \"{3}\".",
-                        count,
-                        name,
-                        _currentLine,
-                        _currentLineText);
-                    throw new ObjParseException(message);
-                }
+                if (!pieces.MoveNext())
+                    goto Fail;
+                piece0 = pieces.Current;
+
+                if (!pieces.MoveNext())
+                    goto Fail;
+                piece1 = pieces.Current;
+
+                if (!pieces.MoveNext())
+                    goto Fail;
+                piece2 = pieces.Current;
+
+                if (!exact || !pieces.MoveNext())
+                    return;
+
+                Fail:
```

Comment:
Please don't use goto or labels in this code.

Reply:
Does this rule apply to the whole class or just the `ExpectPieces` methods?